lib/gnat/consumer_supervisor.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Gnat.ConsumerSupervisor do
+  use GenServer
+  require Logger
+
+  def start_link(settings, options) do
+    GenServer.start_link(__MODULE__, settings, options)
+  end
+
+  def init(settings) do
+    Process.flag(:trap_exit, true)
+    {:ok, task_supervisor_pid} = Task.Supervisor.start_link()
+    connection_name = Map.get(settings, :connection_name)
+    subscription_topics = Map.get(settings, :subscription_topics)
+    consuming_function = Map.get(settings, :consuming_function)
+    send self(), :connect
+    state = %{
+      connection_name: connection_name,
+      connection_pid: nil,
+      consuming_function: consuming_function,
+      status: :disconnected,
+      subscription_topics: subscription_topics,
+      subscriptions: [],
+      task_supervisor_pid: task_supervisor_pid,
+    }
+    {:ok, state}
+  end
+
+  def handle_info(:connect, %{connection_name: name}=state) do
+    case Process.whereis(name) do
+      nil ->
+        Process.send_after(self(), :connect, 2_000)
+        {:noreply, state}
+      connection_pid ->
+        _ref = Process.monitor(connection_pid)
+        subscriptions = Enum.map(state.subscription_topics, fn(topic_and_queue_group) ->
+          topic = Map.fetch!(topic_and_queue_group, :topic)
```

Comment:
OTP supervisors will bail on restarting a child process if it is being restarted too frequently. By default the limit is 3 restarts in 5 seconds will cause the supervisor to fail and will prevent the application from completing the startup process. [These can be tuned here](https://hexdocs.pm/elixir/Supervisor.Spec.html#supervise/2-options)